server/SchemaLoader.js

```diff
@@ -21,56 +21,90 @@ class SchemaLoader extends EventEmitter {
-                    req.onreadystatechange = function() {
-                        if (req.readyState === XMLHttpRequest.DONE) {
-                            // Request was sent, response is downloaded and operation is complete.
-                            if (req.status === 200) {
-                                this.storeSchemas(JSON.parse(req.responseText));
-                            } else if (req.status > 0) {
-                                // Could load the url, but found unexpected status
-                                console.error(
-                                    "Could not fetch schemas: " +
-                                        repo.url +
-                                        endpoint +
-                                        " returned status " +
-                                        req.status
-                                );
-                            }
-                        }
-                    }.bind(this);
-                    req.send();
-                }
+        (repos || [])
+            .filter((repo) => {
+                return !!repo.url;
+            })
+            .forEach((repo) => {
+                this.syncPublicSchemas(repo);
+                this.syncPrivateSchemas(repo);
             });
-        }
         logger.info("Schema syncing done");
         this.emit("schemas-synced");
     }
 
+    syncPublicSchemas(repo) {
+        logger.info(repo.url + ": syncing all public schemas");
+        let endpoint = "/api/schemas/public";
+        this.retrieveSchemas(repo.url, endpoint);
+    }
+
+    syncPrivateSchemas(repo) {
+        if (!repo.apikey) {
+            logger.warn(repo.url + ": missing apikey");
+            return;
+        }
+
+        if (!repo.vendors || !Array.isArray(repo.vendors) || repo.vendors.length === 0) {
+            logger.warn(repo.url + ": vendors not (properly) set");
+            return;
+        }
+
+        logger.info(repo.url + ": syncing " + repo.vendors + " private schemas");
+        let endpoint = "/api/schemas/" + repo.vendors.join(",");
+        this.retrieveSchemas(repo.url, endpoint, repo.apikey);
+    }
+
+    retrieveSchemas(url, endpoint, apikey) {
+        // Also retrieve metadata information, such as
+        // create and update timestamp.
+        let params = "?metadata=1";
+
+        const req = new XMLHttpRequest();
+        req.open("GET", url + endpoint + params);
+
+        if (typeof apikey !== undefined) {
+            req.setRequestHeader("apikey", apikey);
+        }
+
+        req.onreadystatechange = function() {
+            if (req.readyState === XMLHttpRequest.DONE) {
+                // Request was sent, response is downloaded and operation is complete.
+                if ((req.status >= 200 && req.status <= 300) || req.status === 304) {
+                    this.storeSchemas(JSON.parse(req.responseText));
+                } else if (req.status > 0) {
```

Comment:
0 is `UNSENT`, so it seems to me this is vacuously true. (But if you want to keep the check, let's change it to `req.status != XMLHttpRequest.UNSEND`.)

Reply:
I think you're confusing `status` with `readyState` here.

Reply:
Do'h I am. But the question of why to test for `> 0` remains ;)

Reply:
According to the [MSDN docs](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/status):
> The read-only `XMLHttpRequest.status` property returns the numerical status code of the response of the `XMLHttpRequest`. `status` will be an `unsigned short`. Before the request is complete, the value of `status` will be `0`. It is worth noting that browsers report a status of 0 in case of XMLHttpRequest errors too.

Reply:
Aha. So it'd be 0 if the request is `DONE` but failed due to e.g. a networking error, rather than an HTTP protocol-level error. But in that case we'd want to log, too, no? :thinking: